lib/supavisor/monitoring/prom_ex.ex

```diff
@@ -82,6 +82,30 @@ defmodule Supavisor.Monitoring.PromEx do
     metrics
   end
 
+  @spec do_cache_tenants_metrics() :: :ok
+  def do_cache_tenants_metrics() do
+    metrics = get_metrics() |> String.split("\n")
+
+    Registry.select(Supavisor.Registry.TenantSups, [{{:"$1", :_, :_}, [], [:"$1"]}])
+    |> Enum.uniq()
+    |> Enum.each(fn tenant ->
+      filtered =
+        metrics
+        |> Stream.filter(&String.contains?(&1, "tenant=\"#{tenant}\""))
```

Comment:
I believe `Stream` is not going to make much of a difference here, because you are immediately running the stream with one operation. It's fine to keep the `Stream` module, but `Enum` should be more clear here. WDYT?

I see a possible performance issue, because you are traversing the `metrics` list multiple times (the entire list for each tenant). If the list of metrics is not big enough, then this is fine. But otherwise, it would be nice to reduce the list for each tenant interaction (maybe Enum.reduce + Enum.split_with).

Another alternative would be to categorize each line by capturing the tenant name - using regex and named captures -, and then group the lines by tenant. I'm not entirely sure if this way is going to be faster, but could be an alternative here.

WDYT? If the lists are not big enough, you can ignore me :)

Reply:
Good call! Can you please take another look at the updated function?https://github.com/supabase/supavisor/pull/147/commits/452c6e41c5fb7f8de89184c3234c21d7c4a47d2d

Reply:
@philss yeah this concern is because right now our Realtime metrics endpoint is 20MB and we need to add metrics there. Metrics per tenant explode the total count so that list of metrics can get very large very easily.

Reply:
Point is ... whatever you think is fine that was just my concern.

Reply:
@chasers oh, I see. If think this solution is good for now, but may not be good for the long run.

I'm going to investigate if we could query directly PromEx and avoid this filtering in memory, against strings.
For what I could understand, we cannot. But I'm going to double check.

Reply:
another possible solution is to run PromEx per tenant, but it will take some time to implement, and it's unclear how to grab these metrics for all tenants

Reply:
> another possible solution is to run PromEx per tenant, but it will take some time to implement, and it's unclear how to grab these metrics for all tenants at once

Yeah, this could be a good way to solve the problem! I'm not sure if PromEx is adapted for this kind of usage, so it could require modifications to the lib.

> I'm going to investigate if we could query directly PromEx and avoid this filtering in memory, against strings.
For what I could understand, we cannot. But I'm going to double check.

After double checking, I can confirm that we cannot do that unless we use private APIs.

I believe a small performance / memory usage improvement could be to make the [`get_metrics()`](https://github.com/supabase/supavisor/blob/2540d0ad5e80149bde428faa6158b787f6d09531/lib/supavisor/monitoring/prom_ex.ex#L67) function to return a list of metrics instead of the the joined string. This way you would only join once per tenant, and wouldn't need to split like you do on line 85. The idea is to have fewer string creation, because this is expensive in OTP. WDYT?

Reply:
Yeah okay I think we're going to kick this can down the road a little and optimize later. 

Would like to spend some time researching if we can use PromEx per tenant because that seems like it would be nice.